packages/block-library/src/navigation-link/edit.js

```diff
@@ -210,6 +210,79 @@ export const updateNavigationLinkBlockAttributes = (
 	} );
 };
 
+const useIsInvalidLink = ( kind, type, id ) => {
+	const isPostType =
+		'post-type' === kind || 'post' === type || 'page' === type;
+	const hasId = Number.isInteger( id );
+	const postStatus = useSelect(
+		( select ) => {
+			if ( ! isPostType ) {
+				return null;
+			}
+			const { getEntityRecord } = select( coreStore );
+			return getEntityRecord( 'postType', type, id )?.status;
+		},
+		[ isPostType, type, id ]
+	);
+
+	// Check Navigation Link validity if:
+	// 1. Link is 'post-type'.
+	// 2. It has an id.
+	// 3. It's neither null, nor undefined, as valid items might be either of those while loading.
+	// If those conditions are met, check if
+	// 1. The post status is published.
+	// 2. The Navigation Link item has no label.
+	// If either of those is true, invalidate.
+	const isInvalid =
+		isPostType && hasId && postStatus && 'publish' !== postStatus;
+
+	return isInvalid;
+};
+
+const useMissingText = ( type, isInvalid ) => {
+	let missingText = '';
+
+	switch ( type ) {
+		case 'post':
+			/* translators: label for missing post in navigation link block */
+			missingText = __( 'Select a post' );
+			break;
+		case 'page':
+			/* translators: label for missing page in navigation link block */
+			missingText = __( 'Select a page' );
+			break;
+		case 'category':
+			/* translators: label for missing category in navigation link block */
+			missingText = __( 'Select a category' );
+			break;
+		case 'tag':
+			/* translators: label for missing tag in navigation link block */
+			missingText = __( 'Select a tag' );
+			break;
+		default:
+			/* translators: label for missing values in navigation link block */
+			missingText = __( 'Add a link' );
+	}
+
+	return (
+		<>
+			<span>{ missingText }</span>
+			{ isInvalid && (
+				<span className="wp-block-navigation-link__invalid-item">
+					{ ' ' }
+					(
+					{
+						__(
+							'Invalid'
+						) /* translators: Whether or not the navigation link is invalid. */
```

Comment:
I'm not 100% sure the translators comment will be picked up in this position. Usually it's on the line immediately above the string, would be good to confirm if having it after works too.